importer: Add a jsonnet URL importer

camh- opened the pull request:
The importer implements the jsonnet.Importer interface and allows importing
from HTTPS sources as well as local filesystem sources.

A path starting with a double-slash (`//`) is fetched via HTTPS. Paths
starting with a single-slash (`/`) are fetched from the filesystem.

Add tests for the importer.

juliaogris commented on Makefile:
```diff
@@ -0,0 +1,62 @@
+all: lint test check-coverage build  ## build, lint, test and check coverage
+	@[ -e .git/rebase-merge ] && git --no-pager log -1 --pretty='%h %s'
```
`@[ ! -e .git/rebase-merge ] || git --no-pager log -1 --pretty='%h %s'`

camh- replied:
Done. Changed to `if` as discussed in person.

camh- replied:
Done in 6b86995d99b025b7e006e9aca4fd18187b62bc68

juliaogris commented on Makefile:
```diff
@@ -0,0 +1,62 @@
+COLOUR_WHITE  = $(shell tput setaf 7 2>/dev/null)
+
+help:
+	@awk -F ':.*## ' 'NF == 2 && $$1 ~ /^[A-Za-z_-]+$$/ { printf "$(COLOUR_WHITE)%-30s$(COLOUR_NORMAL)%s\n", $$1, $$2}' $(MAKEFILE_LIST) | sort
```
Add digits: `^[A-Za-z0-9_-]+`

camh- replied:
Done

camh- replied:
Done in bde16ab51f54405e2d35a0d96b149cd74c95e972

samuong commented on importer/importer_test.go:
```diff
@@ -0,0 +1,245 @@
+
+func TestAppendSearchFromEnv(t *testing.T) {
+	i := Importer{}
+	os.Setenv("IMPORTER_TEST", "//example.com/path:/local/path")
```
`os.Setenv` (and `os.Unsetenv`) can fail, so `require.NoError`?

camh- replied:
Done in 38a81ffbe14da36daf32f9c5ec5621edda514a80

camh- commented on importer/importer_test.go:
```diff
@@ -0,0 +1,245 @@
+func TestAppendSearchFromEnv(t *testing.T) {
+	i := Importer{}
+	os.Setenv("IMPORTER_TEST", "//example.com/path:/local/path")
+	defer os.Unsetenv("IMPORTER_TEST")
```
I've removed the `defer` altogether. As you say, it is not really restoring the state correctly, and there's no need to do that anyway.

camh- replied:
Done in eadaa2fba52b8c94cb5a48f50f7591e0796db7a8

camh- commented on importer/importer.go:
```diff
@@ -0,0 +1,136 @@
+//
+// Import will cache the result and return it the next time that path is
+// requested.
+func (i *Importer) Import(source, imp string) (jsonnet.Contents, string, error) {
```
Done in 2ac7050dcdeed02c0791109926418fae0ee8d0d6

camh- commented on importer/importer.go:
```diff
@@ -0,0 +1,136 @@
+
+var noContent = jsonnet.Contents{}
+
+type URLFetcher interface {
```
Done in 8fb98aba8e49321f7b312e96eda3da1bd9400f74

camh- commented on importer/importer.go:
```diff
@@ -0,0 +1,136 @@
+	"path/filepath"
+	"strings"
+
+	jsonnet "github.com/google/go-jsonnet"
```
I just copied that from another project. I actually wasn't sure what name it would be imported as, so now I know (https://golang.org/ref/spec#Import_declarations).

Fixed in 6edfa3b3b04bf142e32d98ef862ae58c4a0cda37

juliaogris replied:
Imo if the package name is not the same as the last directory of the import path it should be a named import, otherwise it might be ambiguous...
ie it is not necessarily obvious that `"github.com/google/go-jsonnet"` imports the `jsonnet` package.

camh- replied:
Good point. I actually wondered what the import name would be for the directory `go-jsonnet` and from this comment I learnt that the directory name is not relevant to the naming of the package scope. I think then it is useful to specify that name if it is different to the directory name so it is clear what name is being used from looking at the source code.

I will drop that fixup commit.

camh- commented on importer/importer.go:
```diff
@@ -0,0 +1,136 @@
+		i.cache[imp] = content
+	}
+	return content, err
+}
```
It does last forever (for the life of the importer) and that is required. I've added a link to the `jsonnet.Importer` interface in the godocs, and extended the comment to describe the cache behaviour.

Done in 6e4439a6e7134017c6b0ca93e49ae82eea0b4bf0

camh- commented on importer/importer.go:
```diff
@@ -0,0 +1,136 @@
+	return i.Fetcher
+}
+
+func canonicalise(imp string) string {
```
ack. renamed to `addScheme`

Done in 38c20ef25c05364f1c0a8dfb51912a6c9ad1971b

camh- commented:
@juliaogris I've addressed all your comments, and reworked the makefile a little so I can run lint from docker (I don't have it installed on my laptop). PTAL.

camh- commented:
I've rename the repo from `jsonnet-ext` to `jsonnext`. Thanks for the name @juliaogris.

camh- commented:
@juliaogris I've squashed the fixups and force pushed. I snuck an extra change or two in, so take a look at the forced-push link(s): https://github.com/camh-/jsonnext/compare/518162cea30c999d89390e7183d29b63522f40eb..3caafeec34450bc4a0bb0e7c410e9e392f01bf29

juliaogris commented:
Lgtm, I wonder though why you recreated the go install command rather than used it?

juliaogris commented:
Also the GitHub app interface is not great... Doesn't even say if you need another approval to merge

camh- commented:
> Doesn't even say if you need another approval to merge

That's because I don't. I guess I don't have the option to dismiss approvals on change turned on.

> I wonder though why you recreated the go install command rather than used it?

I probably should not have even put that in. There are no binaries in this repo, but that install target is part of my "standard" go makefile.

However, good point. I have changed it to just use `go install` (and fixed a couple of little related things too).
